src/main/java/com/github/manolo8/darkbot/extensions/plugins/PluginHandler.java

```diff
@@ -139,16 +143,37 @@ private void loadPlugin(Plugin plugin) throws IOException, PluginLoadingExceptio
             if (plJson == null) {
                 throw new PluginLoadingException("The plugin is missing a plugin.json in the jar root", plugin);
             }
-            try (InputStreamReader isr = new InputStreamReader(jar.getInputStream(plJson), StandardCharsets.UTF_8)) {
-                PluginDefinition plDef = GSON.fromJson(isr, (Type) PluginDefinition.class);
-                plugin.setDefinition(plDef);
-            }
+            plugin.setDefinition(readPlugin(jar.getInputStream(plJson)));
             testUnique(plugin);
             testCompatibility(plugin);
             testSignature(plugin, jar);
         }
     }
 
+    private void checkUpdate(Plugin plugin) throws IOException {
+        if (plugin.getDefinition().update == null) return;
+
+        PluginDefinition plDef = readPlugin(plugin.getDefinition().update.openStream());
+        if (plDef.download == null) return;
+        if (plugin.getDefinition().version.compareTo(plDef.version) >= 0) return;
+
+        try (InputStream is = plDef.download.openConnection().getInputStream()) {
+            Files.copy(is, plugin.getFile().toPath(), StandardCopyOption.REPLACE_EXISTING);
+        }
+        Popups.showMessageAsync(I18n.get("plugins.update.title"),
+                I18n.get("plugins.update.message", plugin.getName(),
+                        plugin.getDefinition().version, plDef.version),
+                JOptionPane.INFORMATION_MESSAGE);
+    }
+
+    private PluginDefinition readPlugin(InputStream is) throws IOException {
+        PluginDefinition plDef;
+        try (InputStreamReader isr = new InputStreamReader(is, StandardCharsets.UTF_8)) {
+            plDef = GSON.fromJson(isr, (Type) PluginDefinition.class);
+        }
+        return plDef;
```

Comment:
```suggestion
        try (InputStreamReader isr = new InputStreamReader(is, StandardCharsets.UTF_8)) {
            return GSON.fromJson(isr, (Type) PluginDefinition.class);
        }
```
I belive that is all you need, since any exception will simply not need a return (no catch)